Source/Core/ProofObligationDescription.cs

```diff
@@ -0,0 +1,104 @@
+namespace Microsoft.Boogie;
+
+/// <summary>
+/// A multi-faceted description of a proof obligation. This class is intended
+/// to provide several forms of human-readable text summarizing the meaning of
+/// a proof obligation: one for the case where the proof has succeeded, one
+/// for the case where the proof has failed, and one for the case of quickly
+/// identifying the type of proof obligation (for inclusion in long lists of
+/// proof obligations, for example).
+/// </summary>
+public abstract class ProofObligationDescription {
```

Comment:
Could this be further sub-classed in the future to add more metadata? That could be a good thing (and would make the new field a more future-proof replacement for `ErrorData`). But that might influence the class naming and documentation.

Reply:
That's exactly the idea I have in mind. Currently, Dafny passes in strings for the `ErrorData` field. As things are right now, we could begin creating subclasses for _some_ kinds of Dafny assertions, and leave some as strings, allowing us to incrementally migrate error types without, at first, changing any of the messages. Once we've finished that (which might take a while, because there are a lot of them), all of the messages will be in one place and we can review them for consistency, clarity, etc. Ultimately, any subclasses could have any extra metadata that would be useful (and I expect it will be necessary in some cases, because some messages include things references to source constructs in Dafny).

Reply:
Cool. This does mean that, since the new field is typed to be a `ProofObligationDescription` rather than just an `object`, we are asserting that any tooling built on Boogie should aim to provide such descriptive phrases, and further subclass if they want to add more arbitrary metadata.

Should we deprecate `ErrorData`?

Reply:
Yeah, my intent is that new tooling would do exactly that.

I'd personally like to deprecate `ErrorData`, but it could be useful to get feedback from other folks developing Boogie clients before doing so.